bpo-32121: Add reverse argument to tracemalloc.Traceback.format

Jesse-Bakker opened the pull request:
<!-- issue-number: bpo-32121 -->
https://bugs.python.org/issue32121
<!-- /issue-number -->

Commit:
Added reverse argument to tracemalloc.Traceback.format to allow
reversing the order of the frames in the output

vstinner commented on Lib/tracemalloc.py:
```diff
@@ -209,11 +209,14 @@ def __str__(self):
         if limit is not None and limit < 0:
             return lines
-        for frame in self[:limit]:
+        frame_slice = self[:limit]
```
Hum, I'm not sure that it's useful to truncate before reversing the frames. I would suggest to truncate after reversing.

Jesse-Bakker replied:
The rationale was that the function is designed to format the frames of the `limit` last calls. If truncating after reversing, it will format the first `limit` calls, which I think would be less intuitive and less useful.

vstinner replied:
My bad, you are right :-)

vstinner reviewed (commented):
LGTM, but I proposed to change the default parameter to reverse=True: see discussion at https://bugs.python.org/issue32121

Commit:
Changed according to comments on the issue.
Reversed default sorting of tracemalloc.Traceback. Renamed argument.
Adjusted docs. Allowed negative limit, truncating from the other side.

Commit:
Fix tests for new Traceback implementation

vstinner commented on Lib/tracemalloc.py:
```diff
@@ -171,16 +171,16 @@ def __repr__(self):
         # frames is a tuple of frame tuples: see Frame constructor for the
         # format of a frame tuple
-        self._frames = frames
+        self._frames = tuple(reversed(frames))
```
Please extend the comment to explain that the C module gives frames from the most recent to the oldest, but that ther reverse order is expected in the Python API.

Commit:
Fix whitespace

vstinner commented on Doc/library/tracemalloc.rst:
```diff
@@ -663,13 +663,14 @@ Traceback
    The :attr:`Trace.traceback` attribute is an instance of :class:`Traceback`
    instance.
```
Reversing the order of frames is a major backward incompatible change for this API, please mention it:

```
.. versionchanged:: 3.7
   Frames are now sorted from the oldest to the most recent, instead of most recent to oldest.
```

Moreover, since it's a backward incompatible change, it must be documented in the "Porting" section of Doc/whatsnew/3.7.rst:

https://docs.python.org/dev/whatsnew/3.7.html#changes-in-the-python-api

Commit:
Fix test_warnings.PyWarningsDisplayTests.test_tracemalloc

Commit:
Added docs for backwards incompatible changes (.. versionchanged:: and
Doc/whatsnew)

Commit:
Add comment about reversing frames

Commit:
Fix formatting in docs

Jesse-Bakker commented:
I have made the requested changes; please review again.

vstinner commented on Lib/test/test_warnings/__init__.py:
```diff
@@ -941,10 +941,10 @@ def func():
             {fname}:5: ResourceWarning: unclosed file <...>
               f = None
             Object allocated at (most recent call first):
```
Hum, the message stil says "most recent call first". It seems like you have to patch Lib/warnings.py too.

vstinner commented on Misc/NEWS.d/next/Library/2017-11-24-00-59-12.bpo-32121.ePbmwC.rst:
```diff
@@ -0,0 +1,5 @@
+sorting the frames from oldest to most recent. ``Traceback.format()`` 
+now accepts negative *limit*, truncating the result to the ``abs(limit)`` 
+oldest frames. To get the old behaviour, one can use the new 
+*most_recent_first* argument to ``Traceback.format()``.
```
Would you mind to credit yourself? Add "Patch by Jesse Bakker."

vstinner reviewed (changes requested):
It seems like the warnings code displaying a traceback must be updated for the "most recent frame first" line.

Commit:
Merge branch 'master' into add-tracemalloc-traceback-format-reverse

Commit:
Fix ResourceWarning message and credited self

Jesse-Bakker commented:
I have made the requested changes; please review again.

vstinner reviewed (approved):
LGTM, thanks!
